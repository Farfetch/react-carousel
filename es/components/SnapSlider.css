.container {
    scroll-snap-type: x mandatory;

    /* Older spec implementation (for ios9/10) */
    scroll-snap-destination: 0% 100%;
    scroll-snap-points-x: repeat(100%);

    -webkit-overflow-scrolling: touch; /* For safari */

    & > * {
        scroll-snap-align: center;
    }
}
