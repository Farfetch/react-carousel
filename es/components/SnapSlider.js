function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

import { ScrollSlider, SwipeSlider } from "../containers";
import { checkSnapSupport } from "../utils";
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import SnapStyles from "./SnapSlider.css";
import cx from 'classnames';

var SnapSlider = /*#__PURE__*/function (_Component) {
  _inherits(SnapSlider, _Component);

  var _super = _createSuper(SnapSlider);

  function SnapSlider(props) {
    var _this;

    _classCallCheck(this, SnapSlider);

    _this = _super.call(this, props);
    _this.state = {
      isMounted: false
    };
    return _this;
  }

  _createClass(SnapSlider, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      /* eslint-disable-next-line react/no-did-mount-set-state */
      this.setState({
        isMounted: true
      });
    }
  }, {
    key: "render",
    value: function render() {
      if (!this.state.isMounted || checkSnapSupport()) {
        return /*#__PURE__*/React.createElement(ScrollSlider, _extends({}, this.props, {
          className: cx(this.props.className, SnapStyles.container)
        }));
      }

      var _this$props = this.props,
          limitScroll = _this$props.limitScroll,
          otherProps = _objectWithoutProperties(_this$props, ["limitScroll"]);

      return /*#__PURE__*/React.createElement(SwipeSlider, _extends({}, otherProps, {
        disableSwipe: false,
        hasKeysNavigation: false
      }));
    }
  }]);

  return SnapSlider;
}(Component);

SnapSlider.defaultProps = {
  limitScroll: false
};
SnapSlider.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  limitScroll: PropTypes.bool
};
export default SnapSlider;