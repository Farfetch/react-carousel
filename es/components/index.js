export { default as Arrow } from "./Arrow";
export { default as Bullets } from "./Bullets";
export { default as Carousel } from "./Carousel";
export { default as ScrollSlider } from "./ScrollSlider";
export { default as UnevenItemsScrollSlider } from "./UnevenItemsScrollSlider";
export { default as SnapSlider } from "./SnapSlider";
export { default as SwipeSlider } from "./SwipeSlider";