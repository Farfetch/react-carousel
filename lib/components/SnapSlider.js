"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _containers = require("../containers");

var _utils = require("../utils");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _SnapSlider = _interopRequireDefault(require("./SnapSlider.css"));

var _classnames = _interopRequireDefault(require("classnames"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var SnapSlider = /*#__PURE__*/function (_Component) {
  _inherits(SnapSlider, _Component);

  var _super = _createSuper(SnapSlider);

  function SnapSlider(props) {
    var _this;

    _classCallCheck(this, SnapSlider);

    _this = _super.call(this, props);
    _this.state = {
      isMounted: false
    };
    return _this;
  }

  _createClass(SnapSlider, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      /* eslint-disable-next-line react/no-did-mount-set-state */
      this.setState({
        isMounted: true
      });
    }
  }, {
    key: "render",
    value: function render() {
      if (!this.state.isMounted || (0, _utils.checkSnapSupport)()) {
        return /*#__PURE__*/_react["default"].createElement(_containers.ScrollSlider, _extends({}, this.props, {
          className: (0, _classnames["default"])(this.props.className, _SnapSlider["default"].container)
        }));
      }

      var _this$props = this.props,
          limitScroll = _this$props.limitScroll,
          otherProps = _objectWithoutProperties(_this$props, ["limitScroll"]);

      return /*#__PURE__*/_react["default"].createElement(_containers.SwipeSlider, _extends({}, otherProps, {
        disableSwipe: false,
        hasKeysNavigation: false
      }));
    }
  }]);

  return SnapSlider;
}(_react.Component);

SnapSlider.defaultProps = {
  limitScroll: false
};
SnapSlider.propTypes = {
  children: _propTypes["default"].node.isRequired,
  className: _propTypes["default"].string,
  limitScroll: _propTypes["default"].bool
};
var _default = SnapSlider;
exports["default"] = _default;
module.exports = exports.default;