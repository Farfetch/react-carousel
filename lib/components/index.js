"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Arrow", {
  enumerable: true,
  get: function get() {
    return _Arrow["default"];
  }
});
Object.defineProperty(exports, "Bullets", {
  enumerable: true,
  get: function get() {
    return _Bullets["default"];
  }
});
Object.defineProperty(exports, "Carousel", {
  enumerable: true,
  get: function get() {
    return _Carousel["default"];
  }
});
Object.defineProperty(exports, "ScrollSlider", {
  enumerable: true,
  get: function get() {
    return _ScrollSlider["default"];
  }
});
Object.defineProperty(exports, "UnevenItemsScrollSlider", {
  enumerable: true,
  get: function get() {
    return _UnevenItemsScrollSlider["default"];
  }
});
Object.defineProperty(exports, "SnapSlider", {
  enumerable: true,
  get: function get() {
    return _SnapSlider["default"];
  }
});
Object.defineProperty(exports, "SwipeSlider", {
  enumerable: true,
  get: function get() {
    return _SwipeSlider["default"];
  }
});

var _Arrow = _interopRequireDefault(require("./Arrow"));

var _Bullets = _interopRequireDefault(require("./Bullets"));

var _Carousel = _interopRequireDefault(require("./Carousel"));

var _ScrollSlider = _interopRequireDefault(require("./ScrollSlider"));

var _UnevenItemsScrollSlider = _interopRequireDefault(require("./UnevenItemsScrollSlider"));

var _SnapSlider = _interopRequireDefault(require("./SnapSlider"));

var _SwipeSlider = _interopRequireDefault(require("./SwipeSlider"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }