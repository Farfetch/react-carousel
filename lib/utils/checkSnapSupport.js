"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

function checkSnapSupport() {
  return 'scrollSnapAlign' in document.documentElement.style || 'webkitScrollSnapAlign' in document.documentElement.style || 'msScrollSnapAlign' in document.documentElement.style;
}

var _default = checkSnapSupport;
exports["default"] = _default;
module.exports = exports.default;