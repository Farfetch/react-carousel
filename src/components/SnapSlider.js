import { ScrollSlider } from '#containers';
import { quad } from '#utils';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import cx from 'classnames';

class SnapSlider extends Component {
    constructor(props) {
        super(props);

        this.state = {
            isMounted: false,
        };
    }

    render() {
        return (
            <ScrollSlider
                {...this.props}
                className={cx(this.props.className, 'snapSliderContainer')}
            />
        );
    }
}

SnapSlider.defaultProps = {
    animationDuration: 0.3,
    animationTimingFunction: quad,
};

SnapSlider.propTypes = {
    children: PropTypes.node.isRequired,
    className: PropTypes.string,
    animationDuration: PropTypes.number,
    animationTimingFunction: PropTypes.func,
};

export default SnapSlider;
