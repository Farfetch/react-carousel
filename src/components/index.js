export { default as Arrow } from './Arrow';
export { default as Bullet } from './Bullet';
export { default as Bullets } from './Bullets';
export { default as BulletsInfiniteWrapper } from './BulletsInfiniteWrapper';
export { default as Carousel } from './Carousel';
export { default as SnapSlider } from './SnapSlider';
export { default as ScrollSlider } from './ScrollSlider';
export { default as UnevenItemsScrollSlider } from './UnevenItemsScrollSlider';
export { default as SwipeSlider } from './SwipeSlider';
