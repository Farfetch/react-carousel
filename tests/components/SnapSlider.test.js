import { SnapSlider } from '../../src/components/SnapSlider';
import { createSerializer } from 'enzyme-to-json';
import { shallow } from 'enzyme';
import React from 'react';
import checkSnapSupport from '../../src/utils/checkSnapSupport';

jest.mock('../../src/utils/checkSnapSupport');

beforeAll(() => {
    expect.addSnapshotSerializer(createSerializer());
});

describe('<SnapSlider />', () => {
    test('should render correctly the scroll version on SSR', () => {
        const wrapper = shallow(
            <SnapSlider>
                <div>Test 1</div>
                <div>Test 2</div>
            </SnapSlider>,
            {
                // Ensure this will not mount
                disableLifecycleMethods: true,
            }
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('should render the scroll version if has support and is already mounted', () => {
        checkSnapSupport.mockImplementationOnce(() => true);

        const wrapper = shallow(
            <SnapSlider>
                <div>Test 1</div>
                <div>Test 2</div>
            </SnapSlider>,
            {
                // Ensure this will mount
                disableLifecycleMethods: false,
            }
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('should render the swipe version if has no snap support and is already mounted', () => {
        checkSnapSupport.mockImplementationOnce(() => false);

        const wrapper = shallow(
            <SnapSlider>
                <div>Test 1</div>
                <div>Test 2</div>
            </SnapSlider>,
            {
                // Ensure this will mount
                disableLifecycleMethods: false,
            }
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('should pass custom props correctly to swipe version', () => {
        checkSnapSupport.mockImplementationOnce(() => false);

        const wrapper = shallow(
            <SnapSlider
                data-tst={'test-id'}
                className={'TestTheme'}
                disableSwipe
            >
                <div>Test 1</div>
                <div>Test 2</div>
            </SnapSlider>,
            {
                // Ensure this will mount
                disableLifecycleMethods: false,
            }
        );

        expect(wrapper.prop('data-tst')).toBe('test-id');
        expect(wrapper.prop('className')).toBe('TestTheme');
    });

    test('should pass custom props correctly to scroll version', () => {
        checkSnapSupport.mockImplementationOnce(() => true);

        const wrapper = shallow(
            <SnapSlider
                data-tst={'test-id'}
                className={'TestTheme'}
                disableSwipe
            >
                <div>Test 1</div>
                <div>Test 2</div>
            </SnapSlider>,
            {
                // Ensure this will mount
                disableLifecycleMethods: false,
            }
        );

        expect(wrapper.prop('data-tst')).toBe('test-id');
        expect(wrapper.prop('className')).toBe('TestTheme snapSliderContainer');
    });
});
