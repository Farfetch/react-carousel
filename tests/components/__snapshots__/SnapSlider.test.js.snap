// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SnapSlider /> should render correctly the scroll version on SSR 1`] = `
<ScrollSliderContainer
  className="container"
  limitScroll={false}
>
  <div>
    Test 1
  </div>
  <div>
    Test 2
  </div>
</ScrollSliderContainer>
`;

exports[`<SnapSlider /> should render the scroll version if has support and is already mounted 1`] = `
<ScrollSliderContainer
  className="container"
  limitScroll={false}
>
  <div>
    Test 1
  </div>
  <div>
    Test 2
  </div>
</ScrollSliderContainer>
`;

exports[`<SnapSlider /> should render the swipe version if has no snap support and is already mounted 1`] = `
<SwipeSliderContainer
  disableSwipe={false}
  hasKeysNavigation={false}
>
  <div>
    Test 1
  </div>
  <div>
    Test 2
  </div>
</SwipeSliderContainer>
`;
